File: iosgl/iosgl/bt/res/RoleRes.m
//
//  RoleRes.m
//  iosgl
//
//  Created by zhao on 2/4/2020.
//  Copyright © 2020 zhao. All rights reserved.
//

#import "RoleRes.h"
#import "Vector3D.h"
#import "GL_Header.h"
#import "ByteArray.h"
#import "LoadManager.h"
#import "MeshDataManager.h"

 
@interface RoleRes()
@property(nonatomic,strong)NSString* roleUrl;
@property(nonatomic,strong)NSMutableArray<NSString*>* actionAry;
@property(nonatomic,strong)SuccessBlock fun;
@property(nonatomic,strong)Vector3D* ambientLightColor;
@property(nonatomic,strong)Vector3D* sunLigthColor;
@property(nonatomic,strong)Vector3D* nrmDircet;

@property(nonatomic,assign)float ambientLightIntensity;
@property(nonatomic,assign)float sunLigthIntensity;


@end

@implementation RoleRes


-(void)load:(NSString*)url  fun:(SuccessBlock)fun;
{
    self.fun=fun;
    [[LoadManager default] loadUrl: url type:LoadManager.BYTE_TYPE fun:^(NSString* value) {
        NSDictionary* dic=(NSDictionary*)value;
        NSData* netNsData = [[NSData alloc] initWithContentsOfFile:dic[@"data"]];
       
        [self loadComplete:[[ByteArray alloc]init:netNsData]];
    }];
 
}
-(void)loadComplete:(ByteArray*)byte;
{
    self.byte=byte;
    self.version= [self.byte readInt];
    [self readMesh];
 
}
-(void)readMesh;
{
    RoleRes* this=self;
    this.roleUrl = [this.byte readUTF];
    this.ambientLightColor = [Vector3D alloc];
    this.sunLigthColor = [Vector3D alloc];
    this.nrmDircet = [Vector3D alloc];
    
    
    this.ambientLightColor.x = [this.byte readFloat];
    this.ambientLightColor.y = [this.byte readFloat];
    this.ambientLightColor.z = [this.byte readFloat];
    this.ambientLightIntensity = [this.byte readFloat];
    [this.ambientLightColor scaleBy:this.ambientLightIntensity];

    this.sunLigthColor.x = [this.byte readFloat];
    this.sunLigthColor.y = [this.byte readFloat];
    this.sunLigthColor.z = [this.byte readFloat];
    this.sunLigthIntensity = [this.byte readFloat];
    [this.sunLigthColor scaleBy:this.sunLigthIntensity];

    this.nrmDircet.x = [this.byte readFloat];
    this.nrmDircet.y = [this.byte readFloat];
    this.nrmDircet.z = [this.byte readFloat];
    
    [[MeshDataManager default] readData:self.byte batchNum:this.meshBatchNum url:this.roleUrl version:this.version];
}
 
@end

File: android/SceneWorld001/app/src/main/java/z3d/res/RoleRes.java
package z3d.res;

import android.util.Log;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import z3d.base.ByteArray;

import z3d.base.CallBackFun;
import z3d.base.RoleBackFun;
import z3d.filemodel.MeshDataManager;
import z3d.program.Shader3D;
import z3d.units.AnimManager;
import z3d.units.LoadBackFun;
import z3d.units.LoadManager;
import z3d.vo.Vector3D;

public class RoleRes extends BaseRes {
    public String roleUrl;
    public List<String> actionAry ;
    public int meshBatchNum;
    public Vector3D ambientLightColor;
    public float ambientLightIntensity;
    public Vector3D sunLigthColor;
    public float sunLigthIntensity;
    public Vector3D nrmDircet;
    private  static String TAG="ResManager";
    private CallBackFun callBackFun;
    public void load(String url,  CallBackFun backFun)
    {
        callBackFun=backFun;
        RoleRes that=this;
        LoadManager.getInstance().loadUrl(url,LoadManager.BYTE_TYPE, new LoadBackFun() {
            @Override
            public void bfun(HashMap dic) {
                if(dic!=null){
                    ByteArray temp=(ByteArray)dic.get("byte");
                    _byte=temp;
                    RoleRes._waitArr.add(that);
                }else{
                    Log.d(TAG, "bfun: 角色地址错误");
                }
            }
        },null);
    }
    private static List<RoleRes> _waitArr=new ArrayList<>();
    public static void upDataRoleResWaitIng(){
        while (_waitArr.size()>0){
            RoleRes roleRes=   _waitArr.remove(0);
            roleRes.loadComplete(roleRes.callBackFun);
        }
    }
    public void  loadComplete(  CallBackFun bfun)
    {
        _byte.buffer.position(0);
        version =  _byte.readInt();
        readMesh();
        readAction();
        read(new CallBackFun() {
            @Override
            public void StateChange(boolean State) {
                readNext(bfun);
            }
        });;//readimg
    }
    private void   readNext(CallBackFun bfun) {
        read();
        read();
        Log.d("readNext ->",   version+"===> " );
        bfun.StateChange(true);
    }
    public void readMesh() {
        roleUrl = _byte.readUTF();
        if (version >= 16) { //环境参数
            ambientLightColor = new Vector3D();
            sunLigthColor = new Vector3D();
            nrmDircet = new Vector3D();

            ambientLightColor.x = _byte.readFloat();
            ambientLightColor.y = _byte.readFloat();
            ambientLightColor.z = _byte.readFloat();
            ambientLightIntensity = _byte.readFloat();
            ambientLightColor.scaleBy(ambientLightIntensity);

            sunLigthColor.x = _byte.readFloat();
            sunLigthColor.y = _byte.readFloat();
            sunLigthColor.z = _byte.readFloat();
            sunLigthIntensity = _byte.readFloat();
            sunLigthColor.scaleBy(sunLigthIntensity);

            nrmDircet.x = _byte.readFloat();
            nrmDircet.y = _byte.readFloat();
            nrmDircet.z = _byte.readFloat();
        }
        MeshDataManager.getInstance().readData(_byte, meshBatchNum, roleUrl, version);

    }
    private void readAction() {
        ByteArray $actionByte;
        if (version >= 30) {
            $actionByte = getZipByte(_byte);
        } else {
            $actionByte = _byte;
        }
        actionAry = new ArrayList<>();
        int actionNum = $actionByte.readInt();
        for (int i = 0; i < actionNum; i++) {
            String actionName = $actionByte.readUTF();
            Log.d(actionName, "actionName: ");
            AnimManager.getInstance().readData($actionByte, roleUrl + actionName);
            actionAry.add(actionName);
        }


    }

}
